### ISSUE TITLE
Breaking change in storage plugin in v3.3.0

### ISSUE THREAD

**Issue by mflorence99** [ngxs/store#670]
## I'm submitting a...

<pre><code>
[x] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => https://github.com/ngxs/store/blob/master/CONTRIBUTING.md
[ ] Other... Please describe:
</code></pre>

## Current behavior
The fix for issue #538 changed line 26 of `storage.plugin.ts` to:

```ts
if (typeof val !== 'undefined' && val !== null) {
```

However, on line 61, `getValue` for a key that does not yet exist will return `undefined`. In turn, `setItem` on line 65 will serialize that as the string`'undefined'`:

```ts
if (key !== '@@STATE') {
  val = getValue(nextState, key);
}

try {
  this._engine.setItem(key, options.serialize(val));
```

The effect is that the test on line 26 *passes*, the deserialization of the string `'undefined'` *fails* and the value is set to an empty object.

## Expected behavior
Instead, the value should ultimately be set to its *default* state, as it was in `v3.2.0`.

I think the test on line 26 could be coded as:

```ts
if (val !== 'undefined' && val != null) {
```
 
I have patched it to be so in `node_modules/@ngxs/storage-plugin/fesm5/ngxs-storage-plugin.js` and everything behaves as expected.

Note that `val != null` is equivalent to `val !== undefined && val !== null` as per https://stackoverflow.com/questions/38648087/checking-for-null-or-undefined.

**Comment by splincode** [comment]
PR please

**Comment by mflorence99** [comment]
Sure, if you agree I'm on the right track, I'll do that right away!

**Comment by eranshmil** [comment]
The reason for that PR was to fix custom engine under NativeScript.
If you revert that, you will break for other users.
It should be thoroughly investigated.

**Comment by splincode** [comment]
I think you need to add a test current case

**Comment by mflorence99** [comment]
Sure thing - I had time this morning for a PR for my suggested change but not to dive into the test cases. I’ll do that ASAP after the holiday. Also - my apologies - I do not have the context to properly grok the NativeScript implications.

**Comment by eranshmil** [comment]
Maybe the PR author could help. @MrCroft

**Comment by MrCroft** [comment]
Hi everyone,
Sorry for such a late answer, saw this in the morning but couldn't take time to have a look during the work day. I'm home now.
Let me pull current master branch and see what's happening exactly. I know the issue is to the point, maybe too "to the point" for me 😀 
Also, I'm not sure I understand the desired (or appropriate) solution?

What's strange is that I did test with the build I've done back then, and the application worked.

>Note that val != null is equivalent to val !== undefined && val !== null . 

@mflorence99 I actually didn't know that... what started this was that when working with applicationSettings (mobile development), the first time I would start the app (fresh/clean build) I would get this: `Error occurred while deserializing the store value, falling back to empty object.` and the state wasn't initialized with the defaults I provided in my state class under `defaults: { ... }`.
Then, looking at the getItem implementation:
```
getItem(key: any) {
  return applicationSettings.getString(key);
}
```
I've checked what does `applicationSettings.getString(key)` return for a key that doesn't exist, and it returns `undefined` (as opposed to `null` when using localStorage in the browser). With that in mind, I actually thought that the check for `'undefined'` without `typeof` was a mistake.

Anyway, let me have a look now.

**Comment by eranshmil** [comment]
@MrCroft Thanks for your reply.
A quick patch for all cases would be changing line 61 to:
```typescript
  val = getValue(nextState, key) || undefined;
```

@markwhitfeld What do you think about this solution?

**Comment by mflorence99** [comment]
@eranshmil My apologies for interjecting, but my tests showed that `getValue` for a key that does not yet exist already returns `undefined`. The code on line 65 will serialize that as the string `'undefined'` which will pass the test on 26. 

So in effect that patch would be:

```ts
val = /* getValue(nextState, key) -> */ undefined || undefined;
```

**Comment by eranshmil** [comment]
You're right, I misunderstood the issue.

There are a few changes to try to cover all possible return values from the storage engines.
But I think the best approach is to add to the storage plugin documentation what value should be set for an undefined key.

**Comment by MrCroft** [comment]
@mflorence99 hmm... seems legit.
And as long as `undefined` is serialized as `'undefined'`, then it shouldn't have passed the original `if (val !== 'undefined' && val !== null) {...` check.
However, I've just double-checked and:

1. my **getItem()** implementation (`return applicationSettings.getString(key);`) does return `undefined`
2. right before line 26 (`if (typeof val !== 'undefined' && val !== null) {`) **val** is `undefined` instead of `'undefined'`

Does that mean that there's something going on with the serialization part? And that it was never discovered, since most people are likely to use localStorage (which returns `null`) for custom engines as well (although I'm not that experienced to know why would one write a custom implementation for any other reason except to replace the default storage). But I would guess that this kind of thing (**val** getting to be `undefined` instead of `'undefined'` before line 26) would be caught by tests.

**Comment by markwhitfeld** [comment]
@mflorence99 and @MrCroft I created and merged a PR with the following fix in line 26 to handle both scenarios:
`if (val !== 'undefined' && typeof val !== 'undefined' && val !== null) {`

The @dev npm package should contain this fix, please could you both confirm that this is fixed for you?
Once confirmed I will publish a release.

**Comment by MrCroft** [comment]
@markwhitfeld yup, just checked to make sure. Working fine on my end!
Thank you!

**Comment by mflorence99** [comment]
Yes -- works perfectly! Many thanks @markwhitfeld and @MrCroft, @eranshmil and @splincode for helping out so quickly.

**Comment by splincode** [comment]
Will be 3.3.1
